runtime/parachains/src/paras.rs

```diff
@@ -0,0 +1,1103 @@
+	}
+
+	// note replacement of the code of para with given `id`, which occured in the
+	// context of the given relay-chain block number. provide the replaced code.
+	//
+	// `at` for para-triggered replacement is the block number of the relay-chain
+	// block in whose context the parablock was executed
+	// (i.e. number of `relay_parent` in the receipt)
+	fn note_past_code(
+		id: ParaId,
+		at: T::BlockNumber,
+		now: T::BlockNumber,
+		old_code: ValidationCode,
+	) -> Weight {
+
+		<Self as Store>::PastCodeMeta::mutate(&id, |past_meta| {
+			past_meta.note_replacement(at, now);
+		});
+
+		<Self as Store>::PastCode::insert(&(id, at), old_code);
+
+		// Schedule pruning for this past-code to be removed as soon as it
+		// exits the slashing window.
+		<Self as Store>::PastCodePruning::mutate(|pruning| {
+			let insert_idx = pruning.binary_search_by_key(&at, |&(_, b)| b)
+				.unwrap_or_else(|idx| idx);
+			pruning.insert(insert_idx, (id, now));
+		});
+
+		T::DbWeight::get().reads_writes(2, 3)
+	}
+
+	// does old code pruning.
+	fn do_old_code_pruning(now: T::BlockNumber) -> Weight {
+		let config = configuration::Module::<T>::config();
+		let acceptance_period = config.acceptance_period;
+		if now <= acceptance_period {
+			let weight = T::DbWeight::get().reads_writes(1, 0);
+			return weight;
+		}
+
+		// The height of any changes we no longer should keep around.
+		let pruning_height = now - (acceptance_period + One::one());
+
+		let pruning_tasks_done =
+			<Self as Store>::PastCodePruning::mutate(|pruning_tasks: &mut Vec<(_, T::BlockNumber)>| {
+				let (pruning_tasks_done, pruning_tasks_to_do) = {
+					// find all past code that has just exited the pruning window.
+					let up_to_idx = pruning_tasks.iter()
+						.take_while(|&(_, at)| at <= &pruning_height)
+						.count();
+					(up_to_idx, pruning_tasks.drain(..up_to_idx))
+				};
+
+				for (para_id, _) in pruning_tasks_to_do {
+					let full_deactivate = <Self as Store>::PastCodeMeta::mutate(&para_id, |meta| {
+						for pruned_repl_at in meta.prune_up_to(pruning_height) {
+							<Self as Store>::PastCode::remove(&(para_id, pruned_repl_at));
+						}
+
+						meta.most_recent_change().is_none() && Self::parachain_head(&para_id).is_none()
+					});
+
+					// This parachain has been removed and now the vestigial code
+					// has been removed from the state. clean up meta as well.
+					if full_deactivate {
+						<Self as Store>::PastCodeMeta::remove(&para_id);
+					}
+				}
+
+				pruning_tasks_done as u64
+			});
+
+		// 1 read for the meta for each pruning task, 1 read for the config
+		// 2 writes: updating the meta and pruning the code
+		T::DbWeight::get().reads_writes(1 + pruning_tasks_done, 2 * pruning_tasks_done)
+	}
+
+	/// Schedule a para to be initialized at the start of the next session.
+	#[allow(unused)]
+	pub(crate) fn schedule_para_initialize(id: ParaId, genesis: ParaGenesisArgs) -> Weight {
+		let dup = UpcomingParas::mutate(|v| {
+			match v.binary_search(&id) {
```

Comment:
```suggestion
			if let Err(i) = v.binary_search(&id) {
                        v.insert(i, id);
                        false 
			} else {
			true
			}
```
no idea how to suggest multi line replacements

Reply:
exhaustive match seems more explicit, even though `Result` has only two options.

Reply:
It's not about having two variants, but that only one variants wrapped value is used :) feel free to ignore this, nitpick

Reply:
Since both arms return something, it seems more readable to use `match`. I've replaced the one where the `Ok` arm is a `no-op`, though.